local/man/man3/stats.3
.\" Automatically generated by Pod::Man 2.28 (Pod::Simple 3.31)
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  \*(C+ will
.\" give a nicer C++.  Capital omega is used to do unbreakable dashes and
.\" therefore won't be available.  \*(C` and \*(C' expand to `' in nroff,
.\" nothing in troff, for use with C<>.
.tr \(*W-
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
.    ds C`
.    ds C'
'br\}
.\"
.\" Escape single quotes in literal strings from groff's Unicode transform.
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\"
.\" If the F register is turned on, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.SS), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.\"
.\" Avoid warning from groff about undefined register 'F'.
.de IX
..
.nr rF 0
.if \n(.g .if rF .nr rF 1
.if (\n(rF:(\n(.g==0)) \{
.    if \nF \{
.        de IX
.        tm Index:\\$1\t\\n%\t"\\$2"
..
.        if !\nF==2 \{
.            nr % 0
.            nr F 2
.        \}
.    \}
.\}
.rr rF
.\"
.\" Accent mark definitions (@(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2).
.\" Fear.  Run.  Save yourself.  No user-serviceable parts.
.    \" fudge factors for nroff and troff
.if n \{\
.    ds #H 0
.    ds #V .8m
.    ds #F .3m
.    ds #[ \f1
.    ds #] \fP
.\}
.if t \{\
.    ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.    ds #V .6m
.    ds #F 0
.    ds #[ \&
.    ds #] \&
.\}
.    \" simple accents for nroff and troff
.if n \{\
.    ds ' \&
.    ds ` \&
.    ds ^ \&
.    ds , \&
.    ds ~ ~
.    ds /
.\}
.if t \{\
.    ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.    ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.    ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.    ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.    ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.    ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.\}
.    \" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.    \" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.    \" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.    ds : e
.    ds 8 ss
.    ds o a
.    ds d- d\h'-1'\(ga
.    ds D- D\h'-1'\(hy
.    ds th \o'bp'
.    ds Th \o'LP'
.    ds ae ae
.    ds Ae AE
.\}
.rm #[ #] #H #V #F C
.\" ========================================================================
.\"
.IX Title "stats 3"
.TH stats 3 "2016-02-03" "perl v5.22.1" "User Contributed Perl Documentation"
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.if n .ad l
.nh
.SH "NAME"
stats \- perl library for simple statistics
.SH "SYNOPSIS"
.IX Header "SYNOPSIS"
.Vb 1
\& use stats;
\&
\& $o = new stats;
\& $o\->add (1);
\& $o\->add (2);
\& $o\->add (3);
\& print "N=",$o\->N, "\en";
\& print "mean=",$o\->mean, "\en";
\& print "stdev=",$o\->stdev, "\en";
\& print "min=",$o\->min, "\en";
\& print "max=",$o\->max, "\en";
.Ve
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
The \fBstats\fR library calculates simple statistics in the fashion of a
handheld calculator.  All input and return values are floating point.
.ie n .IP "$s = new stats" 4
.el .IP "\f(CW$s\fR = new stats" 4
.IX Item "$s = new stats"
Construct a new \fBstats\fR object.  It contains all the information
necessary to compute statistics on one set of data.  If you want to
analyze multiple sets of data, create multiple objects.
.ie n .IP "$s\->add (x)" 4
.el .IP "\f(CW$s\fR\->add (x)" 4
.IX Item "$s->add (x)"
Add value \fBx\fR to the set.  This will increase \fBN\fR by 1
.IP "N" 4
.IX Item "N"
Return the number of items in the set.  When constructed, the set
contains no values.
.ie n .IP "$s\->sum" 4
.el .IP "\f(CW$s\fR\->sum" 4
.IX Item "$s->sum"
Return the sum of all the numbers in the set.
.ie n .IP "$s\->mean" 4
.el .IP "\f(CW$s\fR\->mean" 4
.IX Item "$s->mean"
Return the mean value of the set, or \*(L"?\*(R" if the set is empty
.ie n .IP "$s\->stdev" 4
.el .IP "\f(CW$s\fR\->stdev" 4
.IX Item "$s->stdev"
Return the sample standard deviation of the set, or \*(L"?\*(R" if the set
does not have 2 or more items.
.ie n .IP "$s\->min" 4
.el .IP "\f(CW$s\fR\->min" 4
.IX Item "$s->min"
.PD 0
.ie n .IP "$s\->max" 4
.el .IP "\f(CW$s\fR\->max" 4
.IX Item "$s->max"
.ie n .IP "$s\->range" 4
.el .IP "\f(CW$s\fR\->range" 4
.IX Item "$s->range"
.PD
Return the extreme values of the set, or \*(L"?\*(R" if the set is empty.
Range is defined as \fBmax\fR\-\fBmin\fR
.SH "SEE ALSO"
.IX Header "SEE ALSO"
\&\fIstats34\fR\|(3), \fIhisto\fR\|(3), \fIleastsquares\fR\|(3), \fIperm\fR\|(3), \fIsets\fR\|(3), \fInormaldist\fR\|(3),
\&\fIStats\fR\|(3), \fIHisto\fR\|(3)
.SH "AUTHOR"
.IX Header "AUTHOR"
Robert Dowling <\fIRPNCalcN@gmail.com\fR>.


local/lib/stats.pm
package stats;

# Minimal statistics
#
'(C) Copyright 2005-2016 Robert Dowling.';
#
#    This program is free software; you can redistribute it and/or modify
#    it under the terms of the GNU General Public License as published by
#    the Free Software Foundation; either version 2 of the License, or
#    (at your option) any later version.
#
#    This program is distributed in the hope that it will be useful,
#    but WITHOUT ANY WARRANTY; without even the implied warranty of
#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#    GNU General Public License for more details.
#
#    You should have received a copy of the GNU General Public License
#    along with this program; if not, write to the Free Software
#    Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
#
# Usage
#
# $o = new stats;
# $o->add (1);
# $o->add (2);
# $o->add (3);
# print "N=",$o->N, "\n";
# print "mean=",$o->mean, "\n";
# print "stdev=",$o->stdev, "\n";
# print "min=",$o->min, "\n";
# print "max=",$o->max, "\n";

require Exporter;
our @ISA = qw(Exporter); 
# our @EXPORT = qw(new add mean stdev sum N min max);

sub new
{
    my $self = {};
    $self->{N} = 0;
    $self->{SX} = 0;
    $self->{SXX} = 0;
    bless $self;
    $self;
}

sub N ()
{
    my $self = shift;
    $self->{N};
}

sub sum ()
{
    my $self = shift;
    $self->{SX};
}

sub add ($)
{
    my $self = shift;
    my $x = shift;
    $self->{N}++;
    $self->{SX}+=$x;
    $self->{SXX}+=$x*$x;
    $self->{MAX}=$x if $self->{N}==1 || $self->{MAX}<$x;
    $self->{MIN}=$x if $self->{N}==1 || $self->{MIN}>$x;
}

sub mean ()
{
    my $self = shift;
    $self->{N} ? $self->{SX}/$self->{N} : "?";
}

sub stdev ()
{
    my $self = shift;
    $self->{N} > 1 ?
	sqrt (($self->{SXX}-$self->{SX}*$self->{SX}/$self->{N})/($self->{N}-1))
	: "?";
}

sub min ()
{
    my $self = shift;
    $self->{N} ? $self->{MIN} : "?";
}

sub max ()
{
    my $self = shift;
    $self->{N} ? $self->{MAX} : "?";
}

sub range ()
{
    my $self = shift;
    return $self->max - $self->min;
}

1;
__END__

=head1 NAME

stats - perl library for simple statistics

=head1 SYNOPSIS

 use stats;

 $o = new stats;
 $o->add (1);
 $o->add (2);
 $o->add (3);
 print "N=",$o->N, "\n";
 print "mean=",$o->mean, "\n";
 print "stdev=",$o->stdev, "\n";
 print "min=",$o->min, "\n";
 print "max=",$o->max, "\n";

=head1 DESCRIPTION

The B<stats> library calculates simple statistics in the fashion of a
handheld calculator.  All input and return values are floating point.

=over

=item $s = new stats

Construct a new B<stats> object.  It contains all the information
necessary to compute statistics on one set of data.  If you want to
analyze multiple sets of data, create multiple objects.

=item $s->add (x)

Add value B<x> to the set.  This will increase B<N> by 1

=item N

Return the number of items in the set.  When constructed, the set
contains no values.

=item $s->sum

Return the sum of all the numbers in the set.

=item $s->mean

Return the mean value of the set, or "?" if the set is empty

=item $s->stdev

Return the sample standard deviation of the set, or "?" if the set
does not have 2 or more items.

=item $s->min

=item $s->max

=item $s->range

Return the extreme values of the set, or "?" if the set is empty.
Range is defined as B<max>-B<min>

=back

=head1 SEE ALSO

stats34(3), histo(3), leastsquares(3), perm(3), sets(3), normaldist(3),
Stats(3), Histo(3)

=head1 AUTHOR

Robert Dowling <F<RPNCalcN@gmail.com>>.

=cut
